R/calculate_study_day.R

```diff
@@ -0,0 +1,132 @@
+#' calculate_study_day(ae, dm, "AESTDTC", "RFSTDTC", "AESTDY")
+#'
+calculate_study_day <- function(sdtm_in,
+                                dm_domain,
+                                tgdt,
+                                refdt,
+                                study_day_var,
+                                merge_key = "USUBJID") {
+  assertthat::assert_that(is.data.frame(sdtm_in))
+  assertthat::assert_that(is.data.frame(dm_domain))
+  assertthat::assert_that(
+    utils::hasName(dm_domain, refdt),
+    msg = "dm_domain needs to have the variable of refdt."
+  )
+  assertthat::assert_that(
+    utils::hasName(sdtm_in, tgdt),
+    msg = "sdtm_in needs to have the variable of tgdt."
+  )
+  assertthat::assert_that(
+    utils::hasName(dm_domain, merge_key),
+    msg = "dm_domain needs to have the variable of merge_key."
+  )
+  assertthat::assert_that(
+    utils::hasName(sdtm_in, merge_key),
+    msg = "sdtm_in needs to have the variable of merge_key."
+  )
+  assertthat::assert_that(is.character(study_day_var))
+  # check tgdt and study_day_var matching, for example, CMSTDTC matches CMSTDY
+  if (gsub("DTC", "", tgdt, fixed = TRUE) != gsub("DY", "", study_day_var, fixed = TRUE)) {
+    warning(
+      "Target date and the returned study day doesn't match. ",
+      "Expecting matching date and study day, for example, CMENDTC and CMENDY"
+    )
+  }
+
+  original_variables <- names(sdtm_in)
+
+  if (!identical(sdtm_in, dm_domain)) {
+    dm_domain <- unique(dm_domain[c(merge_key, refdt)])
+
+    check_refdt_uniqueness <- dm_domain |>
+      dplyr::group_by(dplyr::pick({{ merge_key }})) |>
+      dplyr::filter(dplyr::n() > 1L)
+    if (nrow(check_refdt_uniqueness) > 0L) {
+      warning(
+        "Reference date is not unique for each patient! ",
+        "Patient without unique reference date will be ingored. ",
+        "NA will be returned for such records."
+      )
+      dm_domain <- dm_domain[
+        !dm_domain[[merge_key]] %in% check_refdt_uniqueness[[merge_key]],
+      ]
+    }
+
+    sdtm_in <- sdtm_in |>
+      dplyr::left_join(
+        dm_domain,
+        by = structure(names = merge_key, .Data = merge_key)
+      )
+  }
+
+  # refdt/tgdt should be in ISO format, otherwise throw warning
+  sdtm_in[[refdt]] <- tryCatch(
+    as.Date(sdtm_in[[refdt]], "%Y-%m-%d"),
+    error = function(e) {
+      warning(
+        "Encountered errors when converting refdt to dates. ",
+        "The warning message is ",
+        e$message,
+        call. = FALSE
+      )
+      NA
+    }
+  )
+  sdtm_in[[tgdt]] <- tryCatch(
+    as.Date(sdtm_in[[tgdt]], "%Y-%m-%d"),
+    error = function(e) {
+      warning(
+        "Encountered errors when converting tgdt to dates. ",
+        "The warning message is ",
+        e$message,
+        call. = FALSE
+      )
+      NA
+    }
+  )
+
+  ref <- sdtm_in[[refdt]]
+  tgt <- sdtm_in[[tgdt]]
+
+  # SDTMIG 4.4.4 Use of the Study Day Variables
+  res <- ifelse(tgt >= ref, tgt - ref + 1L, tgt - ref)
```

Comment:
I would have a dedicated function for this part:

```R
study_day <- compute_study_day(ref, tgt)
```

### `compute_study_day()`

```R
#' Compute study day
#'
#' @description
#' [compute_study_day()] calculates the relative day (study day) of an
#' observation dated `dtc_tar`, starting with the reference date (`dtc_ref`) as
#' day `1`. The study day value is incremented by 1 for each date following the
#' reference date (`dtc_ref`). Dates (`dtc_tar`) prior to reference date
#' (`dtc_ref`) are decreased by 1, with the date preceding `dtc_ref` designated
#' as study day `-1` (there is no study day `0`).
#'
#' The study day implementation follows the definition indicated in the SDTMIG
#' v3.4, Section 4.4.4 _Use of the Study Day Variables_.
#'
#' @param dtc_ref Reference date, a [Date][base::Dates] vector.
#' @param dtc_tar Target date, a [Date][base::Dates] vector.
#'
#' @returns An [integer] vector of relative days, i.e. study days.
#'
#' @examples
#' # When the reference and target dates match the result is 1.
#' sdtm.oak:::compute_study_day(as.Date("2023-01-01"), as.Date("2023-01-01"))
#'
#' # An example spanning two months.
#' sdtm.oak:::compute_study_day(as.Date("2023-01-01"), as.Date("2023-02-11"))
#'
#' # `compute_study_day()` is vectorised over its arguments.
#' ref <- as.Date(c("2023-02-28", "2024-02-28", "2024-02-29", NA_character_))
#' tar <- as.Date(c("2023-03-01", "2024-03-01", "2024-03-01", "2024-03-01"))
#' sdtm.oak:::compute_study_day(ref, tar)
#'
#' # Target dates preceding reference dates result in negative values.
#' sdtm.oak:::compute_study_day(as.Date("2000-09-14"), as.Date("2000-09-13"))
#' sdtm.oak:::compute_study_day(as.Date("1980-06-18"), as.Date("1979-09-07"))
#'
#' @keywords internal
compute_study_day <- function(dtc_ref, dtc_tar) {
  admiraldev::assert_date_vector(dtc_ref)
  admiraldev::assert_date_vector(dtc_tar)

  study_day <-
    ifelse(dtc_tar >= dtc_ref, dtc_tar - dtc_ref + 1L, dtc_tar - dtc_ref)

  as.integer(study_day)
}

```

Reply:
Hi @ramiromagno same as before, this is a small and easy function and I feel like it's an over to write the simple logic into another function if we won't reuse the function for a second time.